### ISSUE TITLE
Scala-cli does not finish running in the case of a scalac StackOverflowError

### ISSUE THREAD

**Issue by jchyb** [VirtusLab/scala-cli#1781]
**Version(s)**
0.1.19

**Describe the bug**
`scala-cli` runs infinitely if it tries compiling code for which scalac throws a StackOverflowError.
Most likely a bloop problem, since with `--server=false` option everything runs as expected (scalac crashes).

**To Reproduce**
```
// main.scala
import scala.deriving.Mirror
import scala.reflect.ClassTag

type TupleUnionLub[T <: Tuple, Lub, Acc <: Lub] <: Lub = T match {
  case (h & Lub) *: t => TupleUnionLub[t, Lub, Acc | h]
  case EmptyTuple     => Acc
}

transparent inline given derived[A](
    using m: Mirror.SumOf[A],
    idClassTag: ClassTag[TupleUnionLub[m.MirroredElemTypes, A, Nothing]]
): Unit = ()

sealed trait Foo
case class FooA(a: Int) extends Foo

val instance = derived[Foo]
```
Run with `scala-cli -S 3.2.1 main.scala`

**Expected behaviour**
scalac should stop running with an error, like when using `--server=false`:

<details>

```scala
Exception in thread "main" java.lang.StackOverflowError
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
        at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
        at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
        at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
(...)
```
</details>

**Comment by lwronski** [comment]
I have one observation, I decrease heap size for bloop and I have the following results:
```
scala-cli bloop exit
scala-cli Hello.scala --bloop-java-opt -Xmx256m
Starting compilation server
Compiling project (Scala 3.2.1, JVM)
java.lang.StackOverflowError
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
	at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
	at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
	at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
	at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
	at dotty.tools.dotc.core.Types$TypeMap.mapOver(Types.scala:5674)
	at dotty.tools.dotc.core.Substituters$.subst(Substituters.scala:25)
	at dotty.tools.dotc.core.Substituters$SubstBindingMap.apply(Substituters.scala:167)
...
```
but it still doesn't finish compiling.

**Comment by lwronski** [comment]
Runnning scala-cli with increased verbosity gives the following results:
```
scala-cli Hello.scala -v -v -v      
...
Compilation cancelled (Scala 3.2.1, JVM)
Received onBuildLogMessage from bloop: LogMessageParams [
  type = WARNING
  task = null
  originId = null
  message = "Disconnecting from deduplication of ongoing compilation for 'project_103be31561-25fccf1679'\nNo progress update for 30 seconds caused bloop to cancel compilation and schedule a new compile.\n                  "
]
Warning: Disconnecting from deduplication of ongoing compilation for 'project_103be31561-25fccf1679'
No progress update for 30 seconds caused bloop to cancel compilation and schedule a new compile.
                  
Received onBuildTaskStart from bloop: TaskStartParams [
  taskId = TaskId [
    id = "2"
    parents = null
  ]
  eventTime = 1674079593628
  message = "Compiling project_103be31561-25fccf1679 (1 Scala source)"
  dataKind = "compile-task"
  data = {"target":{"uri":"file:/private/tmp/scala-demo/.scala-build/?id=project_103be31561-25fccf1679"}}
]
Compiling project (Scala 3.2.1, JVM)
...
```

**Comment by tgodzik** [comment]
I seem to be getting the same error in Bloop. Maybe `StackOverflowError` is not caught by Bloop somewhere? Is it counted as NonFatal?

**Comment by jchyb** [comment]
It is not counted as NonFatal. I will admit I have no idea how bloop works :fearful:, but that doesn't sound like something that would be caught usually

**Comment by tgodzik** [comment]
@jchyb was this reported in the compiler also? We should limit the amount o places where StackOverflowError is thown from the compiler, sicne this can leave the entire Bloop server in a bad state.

**Comment by tgodzik** [comment]
The issue should be fixed by https://github.com/scalacenter/bloop/pull/2001

**Comment by tgodzik** [comment]
We should be able to update Bloop fork now

**Comment by mbesida** [comment]
> The issue should be fixed by [scalacenter/bloop#2001](https://github.com/scalacenter/bloop/pull/2001)

does scala-cli 0.2.0 use version of bloop which has that fix?

**Comment by lwronski** [comment]
> > The issue should be fixed by [scalacenter/bloop#2001](https://github.com/scalacenter/bloop/pull/2001)
> 
> does scala-cli 0.2.0 use version of bloop which has that fix?

Not yet, but it should be included in scala-cli soon
